rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
+
+- Restrict merge access to "GOV.UK Production" [[example](https://github.com/alphagov/govuk-saas-config/pull/55)].
+
+- Tag [relevant Smoke tests](https://github.com/alphagov/smokey/blob/5dc36fc78c12838ec1c8070d6fbc535ae5433b89/docs/writing-tests.md#filtering-by-app) to run for the app [[example](https://github.com/alphagov/smokey/pull/708/commits/91a5ce94b7b96d09f5db87e63fbc3e400788de00)].
+
+- Add a note in the [Release](https://github.com/alphagov/release) app to warn about the change.
+
+- Add a PR template to warn about the change [[example](https://github.com/alphagov/smart-answers/blob/3509ba3e68c62a206874c96111628dd65adae689/.github/pull_request_template.md)].
+
+  - Backfill the PR template as a comment on existing PRs.
+
+- Enable automatic deployments in a Puppet PR [[example](https://github.com/alphagov/govuk-puppet/pull/10590)].
+
+  - Comment on how the app meets the criteria of this RFC.
+
+> **We can be flexible with coverage**. We can defer adding internal repo tests if the code has no impact on the running app e.g. rake tasks [[1](https://github.com/alphagov/email-alert-api/blob/59fc71a58317ef2998f2c0ef102020da3ca9df96/lib/tasks/support.rake)]. But we should not defer removing any unused code.
+
+### Slack notifications
+
+We need to ensure we are notified about a failure in the automation, so that the automation is no worse than the current, manual process. This has already been done for the `Deploy_App_Downstream` job [[1](https://github.com/alphagov/govuk-puppet/pull/10532)] [[2](https://github.com/alphagov/govuk-puppet/pull/10601)].
+
+So that we cover the entire deployment pipeline, we will also update [the `Deploy_App` job](https://github.com/alphagov/govuk-puppet/blob/f57335f6027487b5b2d2205a35ceb411776f09ce/modules/govuk_jenkins/templates/jobs/deploy_app.yaml.erb) to post a Slack message in the #govuk-developers Slack channel, whenever it fails due to an unexpected error.
+
+### Delete [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests)
+
+These sandboxed E2E tests are only run for the apps affected by this RFC, but do not form part of the new safety criteria for automatically deploying them. In other words, they become superfluous, yet their maintenance cost is very high. They can be safely deleted once automatic deployments are enabled for [all of the supported apps](https://github.com/alphagov/publishing-e2e-tests/blob/8412c23c5907a41a3d8b2c9dcd52d4905b139e32/docker-compose.yml).
```

Comment:
Forgive my ignorance, am I right in thinking e2e tests are the only thing that actually test async behaviour across services (aside from production users)? E.g. the new workers are working correctly, I can still publish something, editions still get pulled off a queue and indexed in search.

If that is the case, I wonder if you've already considered how we will test async behaviour without e2e tests?

Reply:
Yes, that's right. Originally the E2E tests were going to feature significantly as part of the safety criteria for enabling CD - that's what [this section](file:///Users/benthorner/govuk/govuk-rfcs/rfc-128-continuous-deployment.md#check-interactions-between-apps-work) is about.

For the reasons in that section, I eventually rejected this approach because:

- No one seems to like our E2E tests, for quite fundamental reasons, so we're unlikely to get very far if we mandate writing a load more of them (no one _wants_ to do that).

- With a bit more thought, I struggle to think of any scenarios that are covered by the E2E tests that aren't (or can't be) covered by [the safety criteria described above](https://github.com/alphagov/govuk-rfcs/blob/continuous-deployment/rfc-128-continuous-deployment.md#safety-concerns).

One thing to really emphasise from that section is that the E2E tests are really artificial: the environment we're running them in has nothing to do with our real environments, which means we get little additional reassurance from them over thorough testing within each app, and contracts for the API interactions between them.

So I think the answer to your last question is: we already do (in part), and the safety criteria in this RFC will start to make that much more comprehensive than it is now (or may ever be if we rely on E2E tests).

This [blog](https://technology.blog.gov.uk/2019/01/29/lessons-learnt-using-contract-testing-in-gov-uk-pay/) is a bit downplayed in the text, but the outcome is basically the same.

Reply:
Thank you, I appreciate you've considered this! Let's delete them.